bigquery_etl/publish_json.py

```diff
@@ -0,0 +1,166 @@
+                sys.exit(1)
+
+            # if it is an incremental query, then the query result needs to be
+            # written to a temporary table to get exported as JSON
+            self._write_results_to_temp_table()
+            self._publish_table_as_json(self.temp_table)
+        else:
+            # for non-incremental queries, the entire destination table is exported
+            result_table = f"{self.dataset}.{self.table}_{self.version}"
+            self._publish_table_as_json(result_table)
+
+    def _publish_table_as_json(self, result_table):
+        """Export the `result_table` data as JSON to Cloud Storage."""
+        prefix = (
+            f"api/{self.api_version}/tables/{self.dataset}/"
+            f"{self.table}/{self.version}/files/"
+        )
+
+        if self.date is not None:
+            # if date exists, then query is incremental and newest results are exported
+            prefix += f"{self.date}/"
+
+        table_ref = self.client.get_table(result_table)
+
+        job_config = bigquery.ExtractJobConfig()
+        job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+
+        # "*" makes sure that files larger than 1GB get split up into JSON files
+        # files are written to a stage directory first
+        destination_uri = f"gs://{self.target_bucket}/" + self.stage_gcs_path + "*.json"
+        extract_job = self.client.extract_table(
+            table_ref, destination_uri, location="US", job_config=job_config
+        )
+        extract_job.result()
+
+        self._gcp_convert_ndjson_to_json(prefix)
+
+    def _gcp_convert_ndjson_to_json(self, gcs_path):
+        """Convert ndjson files on GCP to json files."""
+        blobs = self.storage_client.list_blobs(
+            self.target_bucket, prefix=self.stage_gcs_path
+        )
+        bucket = self.storage_client.bucket(self.target_bucket)
+
+        for blob in blobs:
+            blob_path = f"gs://{self.target_bucket}/{blob.name}"
+            tmp_blob_name = blob_path + ".tmp.gz"
+
+            # stream from GCS
+            with smart_open.open(blob_path) as fin:
+                with smart_open.open(tmp_blob_name, "w") as fout:
+                    fout.write("[\n")
+
+                    for i, line in enumerate(fin):
+                        if i > 0:
+                            fout.write(",\n")
+
+                        fout.write(line.replace("\n", ""))
+
+                    fout.write("]")
+
+        # move all files from stage directory to target directory
+        tmp_blobs = self.storage_client.list_blobs(
+            self.target_bucket, prefix=self.stage_gcs_path
+        )
+
+        for tmp_blob in tmp_blobs:
+            # only copy gzipped files to target directory
+            if "tmp.gz" in tmp_blob.name:
+                # remove .tmp from the final file name
+                file_name = tmp_blob.name.split("/")[-1].replace("tmp.", "")
+                bucket.rename_blob(tmp_blob, gcs_path + file_name)
```

Comment:
It seems like each file has to be moved separately. I tried to rename the entire directory but it didn't work (seems like directories aren't valid blobs)

Reply:
There is still a failure scenario here where we end up with incomplete data in the final bucket. It might be worth it to add some logging in here that will transmit to Airflow so if something happens it's easy to find out on which object.